samcli/lib/utils/stream_writer.py

```diff
@@ -1,28 +1,55 @@
 """
 This class acts like a wrapper around output streams to provide any flexibility with output we need
 """
-from typing import TextIO
+from io import BytesIO, TextIOWrapper
+from typing import Optional, TextIO, Union
 
 
 class StreamWriter:
-    def __init__(self, stream: TextIO, auto_flush: bool = False):
+    def __init__(self, stream: TextIO, stream_bytes: Optional[Union[TextIO, BytesIO]] = None, auto_flush: bool = False):
         """
         Instatiates new StreamWriter to the specified stream
 
         Parameters
         ----------
         stream io.RawIOBase
             Stream to wrap
+        stream_bytes io.TextIO | io.BytesIO
+            Stream to wrap if bytes are being written
         auto_flush bool
             Whether to autoflush the stream upon writing
         """
         self._stream = stream
+        self._stream_bytes = stream_bytes
         self._auto_flush = auto_flush
+        if isinstance(stream, TextIOWrapper):
+            self._stream_bytes = stream
 
     @property
     def stream(self) -> TextIO:
         return self._stream
 
+    def write_bytes(self, output: bytes):
+        """
+        Writes specified text to the underlying stream
+        Parameters
+        ----------
+        output bytes-like object
+            Bytes to write into buffer
+        """
+        # all these ifs are to satisfy the linting/type checking
+        if not self._stream_bytes:
+            return
+        if isinstance(self._stream_bytes, TextIOWrapper):
+            self._stream_bytes.buffer.write(output)
+            if self._auto_flush:
+                self._stream_bytes.flush()
+
+        elif isinstance(self._stream_bytes, BytesIO):
+            self._stream_bytes.write(output)
+            if self._auto_flush:
+                self._stream_bytes.flush()
```

Comment:
```suggestion
        if isinstance(self._stream_bytes, TextIOWrapper):
            self._stream_bytes.buffer.write(output)
        elif isinstance(self._stream_bytes, BytesIO):
            self._stream_bytes.write(output)
            
        if self._auto_flush:
            self._stream_bytes.flush()
```
Pull out the flush check so we only have to write it once?

Reply:
If I understand correctly, since we know `stream_bytes` will be either `TextIO` or `BytesIO`, it should always fall under one of the cases, and thus that auto_flush check will always run?

Reply:
yea i though that also but the linter doesn't seem to recognize that since it gives me this error `error: "object" has no attribute "flush"  [attr-defined]` which I guess is happening because it just assumes its a generic object because of the Union type

Reply:
Interesting, all good then